python/lsst/pipe/base/graph/_implDetails.py

```diff
@@ -40,58 +39,61 @@
+
+    def addProducer(self, key: _T, value: _U):
+        if (existing := self._producers.get(key)) is not None:
+            raise ValueError(f"Only one node is allowed to produce {key}, "
+                             f"the current producer is {existing}")
+        self._producers[key] = value
+        if self._createInverse:
+            self._itemsDict[value].add(key)
+
+    def addConsumer(self, key: _T, value: _U):
+        self._consumers[key].add(value)
+        if self._createInverse:
+            self._itemsDict[value].add(key)
+
+    def getConsumers(self, key: _T) -> set[_U]:
+        return self._consumers.get(key, set())
+
+    def getProducer(self, key: _T) -> Optional[_U]:
+        return self._producers.get(key)
+
+    def getAll(self, key: _T) -> set[_U]:
+        return self.getConsumers(key).union(x for x in (self.getProducer(key),) if x is not None)
+
+    @property
+    def inverse(self) -> Optional[DefaultDict[_U, Set[_T]]]:
+        return self._itemsDict if self._createInverse else None
 
     def makeNetworkXGraph(self) -> nx.DiGraph:
         graph = nx.DiGraph()
-        graph.add_edges_from(self._datasetDictToEdgeIterator())
-        if None in graph.nodes():
-            graph.remove_node(None)
+        for entry in self._producers.keys() | self._consumers.keys():
+            producer = self.getProducer(entry)
+            consumers = self.getConsumers(entry)
+            # This block is for tasks that consume existing inputs
+            if producer is None and consumers:
+                for c in consumers:
+                    graph.add_node(c)
+            # This block is for tasks that produce output that is not consumed
+            # in this graph
+            elif producer is not None and not consumers:
+                graph.add_node(producer)
+            # all other connections
+            else:
+                for consumer in consumers:
+                    graph.add_edge(producer, consumer)
         return graph
 
-    def _datasetDictToEdgeIterator(self) -> Generator[Tuple[Optional[_U], Optional[_U]], None, None]:
-        """Helper function designed to be used in conjunction with
-        `networkx.DiGraph.add_edges_from`. This takes a mapping of keys to
-        `_DatasetTrackers` and yields successive pairs of elements that are to
-        be considered connected by the graph.
-        """
-        for entry in self._container.values():
-            # If there is no inputs and only outputs (likely in test cases or
-            # building inits or something) use None as a Node, that will then
-            # be removed later
-            inputs = entry.inputs or (None,)
-            for inpt in inputs:
-                yield (entry.output, inpt)
-
     def keys(self) -> Set[_T]:
-        return set(self._container.keys())
+        return self._producers.keys() | self._consumers.keys()
+
+    def __contains__(self, value: _T) -> bool:
```

Comment:
I think `__contains__` is looking for a `key` not a `value` (which is what is being checked).  So this is just a comment on the variable name.